metagraph/src/common/threads/threading.hpp

```diff
@@ -60,20 +60,39 @@ class ThreadPool {
     template <class F, typename... Args>
     auto emplace(bool force, F&& f, Args&&... args) {
         using return_type = decltype(f(args...));
+        auto wrap = std::bind(std::forward<F>(f), std::forward<Args>(args)...);
+        auto task_exception = std::make_shared<std::exception_ptr>(nullptr);
         auto task = std::make_shared<std::packaged_task<return_type()>>(
-            std::bind(std::forward<F>(f), std::forward<Args>(args)...)
+            [f=std::move(wrap),task_exception]() mutable -> return_type {
+                try {
+                    return f();
+                } catch (...) {
+                    // catch and store the exception
+                    *task_exception = std::current_exception();
+                    std::rethrow_exception(*task_exception);
```

Comment:
I think we need to remove this one. Otherwise it will throw 2 exceptions: one here and another one later on line 82.

Reply:
There was another issue where even if only one exception was being thrown, it was not getting caught by the main thread. I modified this so that exceptions are now thrown in the main thread.

Reply:
But now there's an issue where exceptions are only handled while the thread pool is joining...

Reply:
Maybe it's fine that it's not caught in the main thread? Why is that a problem?

Reply:
Because if you have a `try...catch` block in the main thread, it won't catch if there's an exception thrown by a worker thread.

Reply:
With the current version, it almost works as expected. The only exception is if `get` is called on all futures and `join` is not called on the `ThreadPool`. In this case, the exception is thrown in the `ThreadPool` destructor, which is undefined behaviour and on my system leads to the exception not being caught.

Reply:
Maybe that's ok that it throws in a different thread? So if you want to catch an exception, you'd need to wrap the task into a try section yourself.

Otherwise, it will wait until join, which is not ideal...

Reply:
There are some uses for having it this way. If we're using the thread pool to process a bunch of independent columns, this this strategy will allow for the ones that ran successfully to still proceed to completion.

The obvious drawback is that if all of the results are needed for a subsequent step, then this approach does a lot of wasteful computation before throwing.

Reply:
I'll have to try again to see if it still works if the user puts a try block in their task, I think it didn't work when I tried it last time.

Reply:
But if a user wants to allow it to continue with one of the tasks failing, he should obviously wrap the task into a try .. catch section.